### PR TITLE
Replace the overridable lookup with an unordered_map

### PR DESCRIPTION
**Pull request by zwoop**

### CODE REVIEW

**Comment by zwoop** [comment]
Two comments on this:

1) We can later maybe incorporate some member offsets, to make _conf_to_memberp() also take advantage of a table like this (but need another key then)

2) Does the order of hash insertions matter at all here? As it is now, it's done in the same order that the original code was in the case statement, i.e. ascending length.

**Comment by SolidWallOfCode** [comment]
1. You're not ready for member offsets in this context.

2. No.

File: src/traffic_server/InkAPI.cc [hunk 1]
```diff
@@ -8428,594 +8431,159 @@ TSHttpTxnConfigStringGet(TSHttpTxn txnp, TSOverridableConfigKey conf, const char
   return TS_SUCCESS;
 }
 
-// This is pretty suboptimal, and should only be used outside the critical path.
+// This is a map of all overridable configurations, with the data type
+static const std::unordered_map<std::string_view, std::tuple<TSOverridableConfigKey, TSRecordDataType>> _overridable_map(
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: This is an invalid use of a reserved name - "all identifiers that ... begin with an underscore are reserved for use as names in the global namespace". This means you can use a leading underscore in a namespace or class, but not at file or global scope.

**Comment by SolidWallOfCode** [comment]
If you want to finesse `_conf_to_memberp`, then you should just remove the data type and replace it with converts or converter pack. Although the basic idea has worked well IMHO, I have liked the style less and less. At this point I think I'd redo it using a protocol class and virtual method overrides. In that case, the data type would be replaced by a converter class instance. For example, entries with data type `INT` would instead have a reference to a file scope converter instance that did integer conversions and took a pointer to int member `int T::*` in the constructor.

**Comment by SolidWallOfCode** [comment]
I know this has been tried before and didn't work, I wonder where that went wrong.

**Comment by zwoop** [comment]
I’ll rename the global, I guess camel case for globals these days?

The comment on incorporating the reverse lookup was for discussion, I don’t know what the best approach is.

Yes, this was tried before, not sure what went wrong, maybe we were not ready for unordered map, and the new goodness we get with 11/14/17 might be difference?

**Comment by SolidWallOfCode** [comment]
I've been using camel plus underscores for globals. E.g. `Overrideable_Map` to be distinct from class names.

File: src/traffic_server/InkAPI.cc [hunk 2]
```diff
@@ -8428,594 +8431,159 @@ TSHttpTxnConfigStringGet(TSHttpTxn txnp, TSOverridableConfigKey conf, const char
...
-    if (!strncmp(name, "proxy.config.http.cache.ignore_accept_language_mismatch", length)) {
-      cnf = TS_CONFIG_HTTP_CACHE_IGNORE_ACCEPT_LANGUAGE_MISMATCH;
-    } else if (!strncmp(name, "proxy.config.http.cache.ignore_accept_encoding_mismatch", length)) {
-      cnf = TS_CONFIG_HTTP_CACHE_IGNORE_ACCEPT_ENCODING_MISMATCH;
-    } else if (!strncmp(name, "proxy.config.http.parent_proxy.connect_attempts_timeout", length)) {
-      cnf = TS_CONFIG_HTTP_PARENT_CONNECT_ATTEMPT_TIMEOUT;
-    }
-    break;
+  if (config != Overridable_Map.end()) {
+    auto &&[c, t] = config->second;
```
Comment (SolidWallOfCode) [inline src/traffic_server/InkAPI.cc]: You might try `std::tie(*conf, *type) = config->second;`.

**Comment by shinrich** [comment]
Requesting that this be pulled into 8.1.x to make cherry-picking of other requested changes less error prone

**Comment by zwoop** [comment]
@shinrich Do we want this for 8.0.x, now that 8.1.x is dead?

**Pull request merged**